evm/src/modules/IntentGateway.sol

```diff
@@ -0,0 +1,609 @@
+     */
+    receive() external payable {}
+
+    /**
+     * @dev Should return the `IsmpHost` address for the current chain.
+     * The `IsmpHost` is an immutable contract that will never change.
+     */
+    function host() public view override returns (address) {
+        return _params.host;
+    }
+
+    /**
+     * @dev Fetch the IntentGateway contract instance for a chain.
+     */
+    function instance(bytes calldata stateMachineId) public view returns (bytes32) {
+        bytes32 gateway = _instances[keccak256(stateMachineId)];
+        return gateway == bytes32(0) ? bytes32(uint256(uint160(address(this)))) : gateway;
+    }
+
+    /**
+     * @dev Checks that the request originates from a known instance of the IntentGateway.
+     */
+    modifier authenticate(PostRequest calldata request) {
+        bytes32 module = bytes32(request.from[:32]);
+        // IntentGateway only accepts incoming assets from itself or known instances
+        if (instance(request.source) != module) revert Unauthorized();
+        _;
+    }
+
+    /**
+     * @notice Sets the parameters for the IntentGateway.
+     * @param p The parameters to be set, encapsulated in a Params struct.
+     */
+    function setParams(Params memory p) public {
+        if (msg.sender != _params.admin) revert Unauthorized();
+        // infinite approval to save on gas
+        IERC20(IDispatcher(p.host).feeToken()).approve(p.host, type(uint256).max);
+
+        p.admin = address(0);
+        _params = p;
+    }
+
+    /**
+     * @notice Retrieves the current parameter settings for the IntentGateway module
+     * @dev Returns a struct containing all configurable parameters
+     * @return Params A struct containing the module's current parameters
+     */
+    function params() public view returns (Params memory) {
+        return _params;
+    }
+
+    /**
+     * @notice Places an order with the given order details.
+     * @dev This function allows users to place an order by providing the order details.
+     * @param order The order details to be placed.
+     */
+    function placeOrder(Order memory order) public payable {
+        address hostAddr = host();
+        // fill out the order preludes
+        order.nonce = _nonce;
+        order.user = bytes32(uint256(uint160(msg.sender)));
+        order.sourceChain = IDispatcher(hostAddr).host();
+
+        bytes32 commitment = keccak256(abi.encode(order));
+
+        // escrow tokens
+        uint256 msgValue = msg.value;
+        uint256 inputsLen = order.inputs.length;
+        for (uint256 i = 0; i < inputsLen; i++) {
+            if (order.inputs[i].amount == 0) revert InvalidInput();
+            address token = address(uint160(uint256(order.inputs[i].token)));
+            if (token == address(0)) {
+                // native token
+                if (msgValue < order.inputs[i].amount) revert InsufficientNativeToken();
+                msgValue -= order.inputs[i].amount;
```

Comment:
Does this line automatically escrow  the native asset?
```suggestion
                msgValue -= order.inputs[i].amount;
```

Reply:
Yes the amount in msg.value has been deposited into the contract